File: 프로그래머스/unrated/150369. 택배 배달과 수거하기/택배 배달과 수거하기.java
class Solution {
    public long solution(int cap, int n, int[] deliveries, int[] pickups) {
        long answer = 0;
        
        int deli = 0;
        int pick = 0;
        
        // 먼 곳부터 탐색
        for(int i=n-1; i>=0; i--) {
            deli += deliveries[i];
            pick += pickups[i];
            
            // deli <= 0 && pick <= 0이면 i번째 집은 안 가도 됨
            // 즉, 예제1 기준 5번 집(i=4)에서 2개 배달 0개 수거 후 cap을 각 빼준다면
            // deli == -2, pick == -4가 되고, 5번까지의 거리를 answer에 추가
            
            // 그 다음 i=3일 때, deli + 1 == -1, pick + 4 == 0
            // 이 경우 5번집 갔다오는 길에 모두 처리 가능하기 때문에 안 가도 됨
            // while문 통과하지 않아서 answer에 추가 안 됨
            
            // 즉 먼 곳부터 탐색해서 deli이나 pick이 양수여서 가야하는 경우만 answer에 추가 
            while(deli > 0 || pick > 0) {
                deli -= cap;
                pick -= cap;
                answer += (i+1)*2;
            }
        }
        
        return answer;
    }
}

File: 프로그래머스/unrated/150369. 택배 배달과 수거하기/택배 배달과 수거하기.py
def solution(cap, n, deliveries, pickups):
    answer = 0
    di = n-1
    pi = n-1
    ds = sum(deliveries)
    ps = sum(pickups)
    
    while ds or ps:
        dist = 0
        
        # 배달
        dcap = cap
        # 먼 곳부터 탐색
        for i in range(di, -1, -1):
            if not deliveries[i]:
                continue
            dist = max(dist, i)
            # 가장 먼 곳까지의 거리 저장
            
            if deliveries[i] <= dcap:
                # 현재 배달 가능 최대 수량보다 적다면
                # 배달(dcap, ds에서 해당 인덱스 값만큼 차감)
                dcap -= deliveries[i]
                ds -= deliveries[i]
                deliveries[i] = 0
            else:
                # 배달 완료 불가
                # 남아있는 dcap만큼만 차감하고, 다음에 i번째 집 가야하니까 di = i 
                deliveries[i] -= dcap
                ds -= dcap
                di = i
                break
                
        # 수거
        pcap = cap
        for i in range(pi, -1, -1):
            if not pickups[i]:
                continue
            dist = max(dist, i)
            
            if pickups[i] <= pcap:
                pcap -= pickups[i]
                ps -= pickups[i]
                pickups[i] = 0
            else:
                pickups[i] -= pcap
                ps -= pcap
                pi = i
                break
        
        if dist >= 0:
            answer += (dist+1)*2 # 왕복
        
    return answer 